feat: add cache-spring module

PhilippHeuer opened the pull request:
- add module `cache-bridge-spring`
- add `XanthicSpringCacheManager`
- allows to use xanthic as spring cache implementation

example usage:
```java
	@Bean
	public CacheManager cacheManager() {
		XanthicSpringCacheManager cacheManager = new XanthicSpringCacheManager(spec -> {
			spec.expiryType(ExpiryType.POST_ACCESS);
		});
                // customize a specific cache
		cacheManager.registerCache("my-custom-cache", spec -> {
			spec.maxSize(10L);
		});

		return cacheManager;
	}
```

not sure if "bridge" is the best word for these kind of integrations.

iProdigy commented on bridge-spring/src/main/java/io/github/xanthic/cache/bridge/spring/XanthicSpringCacheManager.java:
```diff
@@ -0,0 +1,82 @@
+public class XanthicSpringCacheManager implements CacheManager {
+
+	private final Map<String, Cache> cacheMap = new HashMap<>();
+	@Getter
```
@sonatype-lift ignore

iProdigy commented on bridge-spring/src/main/java/io/github/xanthic/cache/bridge/spring/XanthicSpringCache.java:
```diff
@@ -0,0 +1,55 @@
+
+	@Override
+	public <T> T get(@NotNull Object key, @NotNull Callable<T> valueLoader) {
+		return null;
```
Should use `computeIfAbsent` here, but may need some AtomicReference complexity for the exception case

PhilippHeuer replied:
Wanted to do this method last as i was unsure how/when it is used by spring and then forgot about it

iProdigy commented on bridge-spring/src/main/java/io/github/xanthic/cache/bridge/spring/XanthicSpringCacheManager.java:
```diff
@@ -0,0 +1,82 @@
+	 */
+	public void registerCache(String name, Consumer<CacheApiSpec<Object, Object>> spec) {
+		this.cacheMap.put(name, createCache(name, spec));
+		this.customCacheNames.add(name);
```
what's the utility of this set when map keyset exists?

PhilippHeuer replied:
This allows users to check if a cache has been created with the default configuration or if it was customized - we could also turn this into a map to store the spec used to create the cache as well.
This also allows us to check the registerCache method in the test cases fairly easily.

I used the https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/cache/caffeine/CaffeineCacheManager.html as reference for most of the code.

iProdigy commented on bridge-spring/src/main/java/io/github/xanthic/cache/bridge/spring/XanthicSpringCacheManager.java:
```diff
@@ -0,0 +1,82 @@
+	 * @param name the name of the cache
+	 * @param spec configuration for the specified cache
+	 */
+	public void registerCache(String name, Consumer<CacheApiSpec<Object, Object>> spec) {
```
would there be any utility in allowing generic type args for the cache being registered?

PhilippHeuer replied:
I don't think so, the spring `Cache` interface https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/cache/Cache.html uses `Object` everywhere and when using the dynamic cache creation (default way) we also don't know what types are stored in a cache ahead of time.

iProdigy commented on bridge-spring/src/main/java/io/github/xanthic/cache/bridge/spring/XanthicSpringCacheManager.java:
```diff
@@ -0,0 +1,82 @@
+	@Nullable
+	public Cache getCache(@NotNull String name) {
+		return this.cacheMap.computeIfAbsent(name, cacheName -> this.dynamic ? createCache(cacheName, this.spec) : null);
+	}
```
not familiar with spring internals; could this be called concurrently? (requiring map synchronization or impl switch)

PhilippHeuer replied:
Yes, this can be called concurrently when using the `dynamic` mode to create caches on-demand.

iProdigy replied:
gotcha - if `dynamic` can also change randomly, can mark it as volatile to be safe

PhilippHeuer replied:
i don't think that `dynamic` can change dynamically, we just have one method which will disable it with no way to turn it back on. This is sth. which is just set once on app startup.

We could move this option into the constructor to make `dynamic` final.

iProdigy replied:
I think doing `setCacheNames(null)` would enable dynamic mode again, but I don't know what use case would intentionally want to be disabling/enabling this option repeatedly...

PhilippHeuer replied:
ah that is true, but users should choose one mode at startup - changing it after that should not be allowed. so moving it into the constructor might be the best choice.

iProdigy replied:
Agree on final constructor param - easier to maintain & safer for multi threaded apps

iProdigy reviewed (commented):
some Q's (since i'm not a spring expert)

edit: this comment was a mobile fail

iProdigy commented on bridge-spring/src/main/java/io/github/xanthic/cache/bridge/spring/XanthicSpringCache.java:
```diff
@@ -0,0 +1,55 @@
+
+	@Override
+	public void put(@NotNull Object key, Object value) {
+		cache.put(key, value);
```
```suggestion
		cache.put(key, toStoreValue(value));
```

iProdigy commented on bridge-spring/src/main/java/io/github/xanthic/cache/bridge/spring/XanthicSpringCache.java:
```diff
@@ -0,0 +1,55 @@
+
+	@Override
+	protected Object lookup(@NotNull Object key) {
+		return cache.get(key);
```
```suggestion
		return fromStoreValue(cache.get(key));
```

iProdigy replied:
nvm, not needed here https://github.com/spring-projects/spring-framework/blob/4f8516e2c3ca420b1608840ab901bf9df7e4d5f1/spring-context/src/main/java/org/springframework/cache/support/AbstractValueAdaptingCache.java#L58

iProdigy reviewed (commented):
do we need to do this?

PhilippHeuer commented:
> do we need to do this?

yes, these funcs will replace null with `NullValue.INSTANCE` as we don't allow null as cache value and spring will try to cache all results of cachable annotation functions (including null) - which is why we are using `AbstractValueAdaptingCache` as base.

PhilippHeuer commented:
I pushed the following changes:

- rebased onto main
- renamed the module to `cache-spring`
- changed the `.editorconfig` eol from CRLF to LR (all files have LR eol currently)
- should address all open feedback

iProdigy commented on spring/src/main/java/io/github/xanthic/cache/spring/XanthicSpringCacheManager.java:
```diff
@@ -0,0 +1,87 @@
+public class XanthicSpringCacheManager implements CacheManager {
+
+	private final Map<String, Cache> cacheMap = new HashMap<>();
+	@Getter
```
@sonatype-lift ignore

iProdigy commented on spring/src/main/java/io/github/xanthic/cache/spring/XanthicSpringCacheManager.java:
```diff
@@ -0,0 +1,87 @@
+ */
+public class XanthicSpringCacheManager implements CacheManager {
+
+	private final Map<String, Cache> cacheMap = new HashMap<>();
+	@Getter
+	private final Set<String> customCacheNames = new HashSet<>();
```
Use ConcurrentHashMap / ConcurrentHashMap.newKeySet

iProdigy reviewed (commented):
Nice work! just a final nitpick

iProdigy reviewed (approved):
Great to have this integration
